src/libfetchers/cache.cc

```diff
@@ -118,4 +132,17 @@ ref<Cache> getCache()
     return ref<Cache>(cache);
 }
 
+struct disabler {
+    disabler() {
+        getCache()->cacheLookupsEnabled(false);
+    }
+    ~disabler() {
+        getCache()->cacheLookupsEnabled(true);
+    }
+};
+
+std::shared_ptr<disabler> disableCacheLookups(void) {
```

Comment:
This is a fairly standard C++ idiom for providing a safe context region of operation.  By using a shared pointer to a context object it ensures that the object's destructor is called to revert/shutdown the context when corresponding scope is no longer active, regardless of how that scope was exited (return, exception, etc.).  In fact, the `disableCacheLookups` is the expected API point; the `struct disabler` is not intended to be used directly.

Reply:
Yes I know the idiom, I just don't see what the shared pointer gives, compared to directly calling the constructor (like it's done in other places in the repo (the [Finally](https://github.com/NixOS/nix/blob/57b935a8931fb59cdfe521357e76326d467a9a86/src/libutil/finally.hh) or Activity classes for example).

(I apologize in case I miss something obvious. I'm by no means a c++ expert, it just seems to me that these lines are not in line with how things are done in the rest of the codebase)

Reply:
The shared pointer immunizes the operation against caller mis-use such as copying/duplicating, passing by value, etc.

The `Finally` requires the _caller_ to explicitly specify the cleanup activity at each usage point, whereas the shared pointer provides encapsulation and avoids repetition.  

I also have concerns that the use of `Finally` without assignment might be subject to either outright removal or re-ordering of the destructor call due to compiler optimization passes because of the lack of any local reference in the calling location; if `Finally` were assigned to a local variable, that would mitigate my concern, but that's not the current usage pattern and is also outside the scope of this PR.

I'm not sure how `Activity` applies here since I'm not intending any associated `Logger` functionality.

I used the shared pointer idiom here because it is simple, safe, and a relatively common idiom for C++ developers.